### ISSUE TITLE
PrincipalSearcher returns computer objects from AD for UserPrincipal subclasses

### ISSUE THREAD

**Issue by NetMage** [dotnet/runtime#40102]
### Description

Documentation for using QBE with `AccountManagement` recommends creating a `UserPrincipal` subclass when needing to query on non-standard directory attributes.

The code in `PushFilterToNativeSearcher` that converts the Example class into an LDAP query uses `GetObjectClassPortion` to determine the base query for the Example.

When the Example class is not a standard built-in `Principal` subclass, the code calls `ReadStructuralObjectClass` which looks for a `DirectoryObjectClassAttribute` on the subclass to use as a filter.

Unfortunately since the AD schema specifies that Computer inherits from User, the `objectClass` is `user` for both types of objects, and the `objectCategory` should be used for filtering. `GetObjectClassPortion` recognizes this for the standard `UserPrincipal` class and specifies an LDAP filter of `(&(objectCategory=user)(objectClass=user))` (would possibly be better as `objectCategory=Person`), but this does not apply to subclasses.

I suggest the `GetObjectClassPortion` should test for subclasses of the standard classes instead of a direct type comparison on the standard classes so that e.g. using a subclass of `UserPrincipal` will work as expected.

Alternatively, a class attribute to specify `objectCategory` on the subclass that will be used by `ReadStructuralObjectClass` (perhaps becomes `ReadStructuralObjectFilter`?) will allow custom attribute filtering with the expected base query.

This can be worked around (and there are many web articles) by setting the `objectCategory` inside your sub-class, or doing other similar things, but there is no apparent consensus on the best way to handle this, and it probably should just do the right thing.

**Comment by danmosemsft** [comment]
I'm not familiar with this code, but I assume the change you're asking for is here:
https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/System.DirectoryServices.AccountManagement/src/System/DirectoryServices/AccountManagement/AD/ADStoreCtx_Query.cs#L192

What subclasses of UserPrincipal are there? I assume they are in your code? Could you give a free-standing code sample?

**Comment by NetMage** [comment]
Yes, that is the code to change. This article is old, but the section on Extensibility Model at the end covers extending UserPrincipal and still holds: [Managing Directory Security Principals in the .NET Framework 3.5](https://docs.microsoft.com/en-us/archive/msdn-magazine/2008/january/managing-directory-security-principals-in-the-net-framework-3-5)

This [Stack Overflow question](https://stackoverflow.com/q/52090273/2557128) is an example that illustrates the same issue I ran into.

**Comment by NetMage** [comment]
@danmoseley Did you ever look at the sample code on Stack Overflow?
